virttest/vt_cluster/selector.py

```diff
@@ -0,0 +1,302 @@
+    def _gt(left, right):
+        return operator.gt(left, right)
+
+    @staticmethod
+    def _eq(left, right):
+        return operator.eq(left, right)
+
+    @staticmethod
+    def _ne(left, right):
+        return operator.ne(left, right)
+
+    @staticmethod
+    def _contains(left, right):
+        if isinstance(left, list) and isinstance(right, list):
+            return all(item in left for item in right)
+        return operator.contains(left, right)
+
+    @staticmethod
+    def _not_contains(left, right):
+        return not _Operator._contains(left, right)
+
+
+class _Selector(object):
+    """
+    A generic selector engine that matches a target against a set of rules.
+
+    :param selectors: A list of selector dictionaries, where each dictionary
+                      defines a matching rule.
+    :type selectors: list
+    """
+
+    def __init__(self, selectors):
+        self._selectors = selectors
+        self._match_expressions = []
+        for selector in self._selectors:
+            self._match_expressions.append(self._build_expression(selector))
+
+    def _build_expression(self, selector):
+        """
+        Builds a _MatchExpression from a selector dict.
+
+        :param selector: A dictionary defining the matching rule.
+        :type selector: dict
+        :return: A match expression object.
+        :rtype: _MatchExpression
+        """
+        return _MatchExpression(
+            selector.get("key"),
+            selector.get("operator"),
+            selector.get("values"),
+        )
+
+    def _get_value_from_target(self, target, key):
+        """
+        Retrieves a value from the target object based on a key.
+
+        This method must be implemented by subclasses to define how values
+        are extracted from the target object.
+
+        :param target: The object to extract the value from.
+        :type target: any
+        :param key: The key or attribute name to look up.
+        :type key: str
+        :raises KeyError: If the key is not found in the target.
+        :return: The value corresponding to the key.
+        :rtype: any
+        """
+        if key not in target:
+            raise KeyError(key)
+        return target[key]
+
+    def match(self, target):
+        """
+        Checks if the given target object matches all selector expressions.
+
+        :param target: The object to check.
+        :type target: any
+        :return: `True` if the target matches all rules, `False` otherwise.
+        :rtype: bool
+        """
+        for match_expression in self._match_expressions:
+            try:
+                target_value = self._get_value_from_target(target, match_expression.key)
+            except KeyError:
+                return False
+
+            if not _Operator.operate(
+                match_expression.operator, target_value, match_expression.values
+            ):
+                return False
+        return True
+
+
+class _NodeSelector(_Selector):
```

Comment:
what about moving the _NodeSelector to a new module like "node_selector.py"? The reason I am asking for it is I also need a pool_selector for the resource pools, the PoolSelector class also derive from the _Selector class.

Another choice is I can also add the PoolSelector here, which do you think is better?